iib/workers/tasks/build_merge_index_image.py

```diff
@@ -0,0 +1,278 @@
+    :param int request_id: the ID of the IIB build request.
+    :param str arch: the architecture to build this image for.
+    :param str ocp_version: ocp version which will be added as a label to the image.
+    :param str overwrite_target_index_token: the token used for overwriting the input
+        ``source_from_index`` image. This is required for non-privileged users to use
+        ``overwrite_target_index``. The format of the token must be in the format "user:password".
+    :return: bundles which were added to the index image.
+    :rtype: list
+    """
+    set_request_state(request_id, 'in_progress', 'Adding bundles missing in source index image')
+    log.info('Adding bundles from target index image which are missing from source index image')
+    missing_bundles = []
+    source_bundle_digests = []
+    target_bundle_digests = []
+
+    for bundle in source_index_bundles:
+        if '@sha256:' in bundle['bundlePath']:
+            source_bundle_digests.append(bundle['bundlePath'].split('@sha256:')[-1])
+        else:
+            raise IIBError(
+                f'Bundle {bundle["bundlePath"]} in the source index image is not defined via digest'
+            )
+    for bundle in target_index_bundles:
+        if '@sha256:' in bundle['bundlePath']:
+            target_bundle_digests.append((bundle['bundlePath'].split('@sha256:')[-1], bundle))
+        else:
+            raise IIBError(
+                f'Bundle {bundle["bundlePath"]} in the target index image is not defined via digest'
+            )
+
+    for target_bundle_digest, bundle in target_bundle_digests:
+        if target_bundle_digest not in source_bundle_digests:
+            missing_bundles.append(bundle)
+
+    missing_bundle_paths = [bundle['bundlePath'] for bundle in missing_bundles]
+    if missing_bundle_paths:
+        log.info('%s bundles are missing in the source index image.', len(missing_bundle_paths))
+    else:
+        log.info(
+            'No bundles are missing in the source index image. However, the index image is '
+            'still being rebuilt with the new binary image.'
+        )
+
+    _opm_index_add(
+        base_dir,
+        missing_bundle_paths,
+        binary_image,
+        source_from_index,
+        overwrite_target_index_token,
+    )
+    _add_label_to_index(
+        'com.redhat.index.delivery.version', ocp_version, base_dir, 'index.Dockerfile',
+    )
+    _build_image(base_dir, 'index.Dockerfile', request_id, arch)
+    _push_image(request_id, arch)
+    _create_and_push_manifest_list(request_id, [arch])
+    log.info('New index image created')
+
+    return missing_bundles
+
+
+def _get_bundles_from_deprecation_list(bundles, deprecation_list):
+    """
+    Get a list of to-be-deprecated bundles based on the data from the deprecation list.
+
+    :param list bundles: list of bundles to apply the filter on.
+    :param list deprecation_list: list of deprecated bundle pull specifications (digest).
```

Comment:
It looks like non-digests are supported in the code. Is that intentional?

Reply:
yup, if the bundle is specified by tag, IIB automatically resolves it to its digest.

Reply:
Can you remove the `(digest)` part in the docstring then?

Reply:
ah, I just realized the code is using `_get_resolved_image`. We should be using `_get_resolved_bundles` instead to get the manifest digest and not the manifest list digest. I'll make the change